R/ordering.R

```diff
@@ -82,3 +82,89 @@ setMethod("ordering<-", "DatasetCatalog", function (x, value) {
+#         after <- urls(after)
+#     }
+#     if (!inherits(value, "OrderGroup")) {
+#         value <- urls(value)
+#     }
+#     whi <- ifelse(inherits(after, "OrderGroup"), which(names(x) %in% name(after)), which(entities(x) %in% after))
+#     if ((inherits(value, "OrderGroup") && !name(value) %in% names(x)) || (!inherits(value, "OrderGroup") && !value %in% entities(x))) entities(x) <- c(entities(x), value)
+#     whi2 <- ifelse(inherits(value, "OrderGroup"), which(names(x) %in% name(value)), which(entities(x) %in% value))
+#     entities(x) <- entities(x)[c(setdiff(1:whi, whi2), whi2, setdiff((whi+1):length(entities(x)), whi2))]
+#     return(x)
+# }
+
+#' Copy the order from one dataset to another.
+#'
+#' @param source the dataset you want to copy the order from
+#' @param target the dataset you want to copy the order to
+#' @return returns an object of class [`VariableOrder`] (which can be assigned
+#' to a dataset with [`ordering`])
+#' @examples
+#' \dontrun{
+#' ordering(ds) <- copyOrder(ds1, ds)
+#' }
+#' @export
+copyOrder <- function (source, target) {
+    if (!is.dataset(source) | !is.dataset(target)) {
+        halt("Both source and target must be Crunch datasets.")
+    }
+
+    ord <- entities(ordering(source))
+
+    # make url and alias maps
+    url_to_alias_source <- as.list(structure(aliases(allVariables(source)), .Names=urls(allVariables(source))))
+    alias_to_url_target <- as.list(structure(urls(allVariables(target)), .Names=aliases(allVariables(target))))
+
+    new_ord <- lapply(ord, copyOrderGroup,
+                      source_map = url_to_alias_source,
+                      target_map = alias_to_url_target)
+
+    # drop any null entities, those that were not found in target but in source
+    new_ord <- removeMissingEntities(new_ord)
+    new_ord <- do.call(VariableOrder, new_ord)
+
+    # set catalog URL so show methods work on the new ordering
+    new_ord@catalog_url <- variableCatalogURL(target)
+
+    return(new_ord)
+}
+
+#' Copy the order of a `VariableGroup` (or individual variable URL) from `VariableOrder`
+#'
+#' @param group the group or variable URL to be copied
+#' @param source_map url to alias map for source variables
+#' @param target_map alias to url map for target variables
+#' @return returns either a [`VariableGroup`] (if a group is supplied) or a URL (if just a variable URL is supplied)
+#' @keywords internal
+copyOrderGroup <- function (group, source_map, target_map) {
+    # if there is a single element in group, and it is a character,
+    # just return the URL in the target.
+    if (length(group) == 1 & is.character(group)) {
+        return(target_map[[source_map[[group]]]] %||% NA)
+    }
+
+    # there are groups, so recurse
+    ents <- lapply(entities(group), copyOrderGroup,
+                           source_map = source_map, target_map = target_map)
+
+    # drop any na entities - those that were not found in target but in source
+    ents <- removeMissingEntities(ents)
```

Comment:
I think if you make L158 return `NA_character_`, you can remove this line and L168 will succeed. Then you only have to call this recursive function once, in `copyOrder`, instead of once there and then once again in each group.